### PR TITLE
Zipformer for Common Voice

### PR DESCRIPTION
**Pull request by yfyeung**
## Data Preparation
The `prepare.sh` prepares the English (en) dataset of version 13.0 by default.
| Version | Date | Size | Recorded Hours | Validated Hours | License | Number of Voices | Audio Format |
|-- | -- | -- | -- | -- | -- | -- | --|
|Common Voice Corpus 13.0 | 3/15/2023 | 76.39 GB | 3,209 | 2,429 | CC-0 | 86,942 | MP3|

| Stage | Time | Comment | 
| -- | -- | -- | 
| Stage 0: Download data | - | Use machine in America to download the dataset manuallly |
| Stage 1: Prepare CommonVoice manifest | 50 minutes | num_jobs=8 |
| Stage 2: Prepare musan manifest | Very little | - |
| Stage 3: Preprocess CommonVoice manifest | 5 minutes | - |
| Stage 4: Compute fbank for dev and test subsets of CommonVoice | 2 minutes | 1 GPU |
| Stage 5: Split train subset into 1000 pieces | 10 minutes  | - |
| Stage 6: Compute features for train subset of CommonVoice | 3 hours  | 4 processes with 2 GPUs |
| Stage 7: Combine features for train | 15 minutes  | - |
| Stage 8: Compute fbank for musan | 15 minutes  | - |
| Stage 9: Prepare BPE based lang | 5 minutes  | - |

## Result
|                      |  Dev  | Test  |
|----------------------|-------|-------|
|    greedy search     | 9.96  | 12.54 |
| modified beam search | 9.86  | 12.48 |

To reproduce the above result, use the following commands for training:

```bash
export CUDA_VISIBLE_DEVICES="0,1,2,3"
./pruned_transducer_stateless7/train.py \
  --world-size 4 \
  --num-epochs 30 \
  --start-epoch 1 \
  --use-fp16 1 \
  --exp-dir pruned_transducer_stateless7/exp \
  --max-duration 550
```

and the following commands for decoding:

```bash
# greedy search
./pruned_transducer_stateless7/decode.py \
  --epoch 30 \
  --avg 5 \
  --decoding-method greedy_search \
  --exp-dir pruned_transducer_stateless7/exp \
  --bpe-model data/lang_bpe_500/bpe.model \
  --max-duration 600

# modified beam search
./pruned_transducer_stateless7/decode.py \
  --epoch 30 \
  --avg 5 \
  --decoding-method modified_beam_search \
  --beam-size 4 \
  --exp-dir pruned_transducer_stateless7/exp \
  --bpe-model data/lang_bpe_500/bpe.model \
  --max-duration 600
```

Pretrained model is available at
<https://huggingface.co/yfyeung/icefall-asr-cv-corpus-13.0-2023-03-09-en-pruned-transducer-stateless7-2023-04-17>

The tensorboard log for training is available at
<https://tensorboard.dev/experiment/j4pJQty6RMOkMJtRySREKw/>

### CODE REVIEW

**Comment by desh2608** [comment]
BTW it seems that `num_jobs` > 1 is currently not supported for the CommonVoice Lhotse recipe (see [here](https://github.com/lhotse-speech/lhotse/blob/9b60b51f82f6543024d7090f2af91c0ef2a6e8cd/lhotse/recipes/commonvoice.py#L151)). It may be worth implementing this option to speed up the manifest creation time.

**Comment by csukuangfj** [comment]
> BTW it seems that `num_jobs` > 1 is currently not supported for the CommonVoice Lhotse recipe (see [here](https://github.com/lhotse-speech/lhotse/blob/9b60b51f82f6543024d7090f2af91c0ef2a6e8cd/lhotse/recipes/commonvoice.py#L151)). It may be worth implementing this option to speed up the manifest creation time.

There's an ongoing PR about that 

https://github.com/lhotse-speech/lhotse/pull/1025

File: egs/commonvoice/ASR/local/compute_fbank_commonvoice_splits.py [hunk 1]
```diff
@@ -90,7 +90,7 @@ def compute_fbank_commonvoice_splits(args):
     subset = "train"
```
Comment (csukuangfj) [inline egs/commonvoice/ASR/local/compute_fbank_commonvoice_splits.py:90]: Could you please add a `RESULTS.md` to document the results, pre-trained models, tensorboard logs?

Comment (yfyeung) [reply egs/commonvoice/ASR/local/compute_fbank_commonvoice_splits.py:90]: Sure

File: egs/commonvoice/ASR/RESULTS.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
+## Results
+### GigaSpeech BPE training results (Pruned Stateless Transducer 7)
+
+#### [pruned_transducer_stateless7](./pruned_transducer_stateless7)
+
+See k2-fsa#672 for more details.
```
Comment (csukuangfj) [inline egs/commonvoice/ASR/RESULTS.md:6]:
```suggestion
See #997  for more details.
```

File: egs/commonvoice/ASR/pruned_transducer_stateless7/decoder.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
+# Copyright    2021  Xiaomi Corp.        (authors: Fangjun Kuang)
```
Comment (csukuangfj) [inline egs/commonvoice/ASR/pruned_transducer_stateless7/decoder.py:1]: Could you replace it with a symlink?

Comment (csukuangfj) [reply egs/commonvoice/ASR/pruned_transducer_stateless7/decoder.py:1]: Also for other files like `joiner.py` and `model.py`.

Comment (yfyeung) [reply egs/commonvoice/ASR/pruned_transducer_stateless7/decoder.py:1]: Sure.

**Review by csukuangfj — APPROVED** [review]
Thanks!

**Pull request merged**